gateway/api.go

```diff
@@ -245,14 +245,51 @@ func (gw *Gateway) applyPoliciesAndSave(keyName string, session *user.SessionSta
 		return err
 	}
 
-	lifetime := session.Lifetime(spec.GetSessionLifetimeRespectsKeyExpiration(), spec.SessionLifetime, gw.GetConfig().ForceGlobalSessionLifetime, gw.GetConfig().GlobalSessionLifetime)
+	// calculate lifetime considering access rights
+	lifetime := gw.ApplyLifetime(session, spec)
 	if err := gw.GlobalSessionManager.UpdateSession(keyName, session, lifetime, isHashed); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// GetApiSpecsFromAccessRights from the session.AccessRights returns the collection of api specs
+func (gw *Gateway) GetApiSpecsFromAccessRights(sess *user.SessionState) []*APISpec {
+	var apis []*APISpec
+	if sess != nil && len(sess.AccessRights) > 0 {
+		for apiID := range sess.AccessRights {
+			spec := gw.getApiSpec(apiID)
+			if spec != nil {
+				apis = append(apis, spec)
+			}
+		}
+	}
+
+	return apis
+}
+
+// ApplyLifetime calculates the lifetime for the key. It considers the access rights and the bigger lifetime will be used
+func (gw *Gateway) ApplyLifetime(sess *user.SessionState, specs ...*APISpec) int64 {
```

Comment:
@sredxny what are the risks of moving this logic inside `session.Lifetime`? it makes more sense if this is applied every time we try to get the session Lifetime..

Reply:
@tbuchaillot I would not do so, I tought about it as well. But this logic depends too much on apidefinitions and sometimes in a collection of apidefinitions, that's why I call `gw.GetApiSpecsFromAccessRights(sess)` so, I cannot just move this into the `user` package  because I would get a cycling import issue, so it will need to be refactored and receive all the values (only primitive data types and session), it will make the code less readable because everytime that you want to call the applyLifetime function then you will need to prepare the data to be sent. 
For sure it can be done, but I dont think the result will look beauty

Reply:
@sredxny alright, thanks for explaining! Please create a ticket for that refactor so someone can have a look deeply.